pennylane/resources/factorization.py

```diff
@@ -0,0 +1,159 @@
+
+        and
+
+        .. math::
+
+            h_{pqrs} = \int \frac{\phi_p(r_1)^* \phi_q(r_2)^* \phi_r(r_2) \phi_s(r_1)}{|r_1 - r_2|}
+            dr_1 dr_2.
+
+        The two-electron integrals can be rearranged in the so-called chemist notation which gives
+
+        .. math::
+
+            V_{pqrs} = \int \frac{\phi_p(r_1)^* \phi_q(r_1)^* \phi_r(r_2) \phi_s(r_2)}{|r_1 - r_2|}
+            dr_1 dr_2,
+
+        and the molecular Hamiltonian can be rewritten as
+
+        .. math::
+
+            H = \sum_{\alpha \in \{\uparrow, \downarrow \} } \sum_{pq} T_{pq} a_{p,\alpha}^{\dagger}
+            a_{q, \alpha} + \frac{1}{2} \sum_{\alpha, \beta \in \{\uparrow, \downarrow \} } \sum_{pqrs}
+            V_{pqrs} a_{p, \alpha}^{\dagger} a_{q, \alpha} a_{r, \beta}^{\dagger} a_{s, \beta}.
+
+        with
+
+        .. math::
+
+            T_{pq} = h_{pq} - \frac{1}{2} \sum_s h_{pssq}.
+
+
+        This notation allows a low-rank factorization of the two-electron integral. The objective of
+        the factorization is to find a set of symmetric matrices, :math:`L`, such that
+
+        .. math::
+
+               V_{ijkl} = \sum_r^R L_{ij}^{(r)} L_{kl}^{(r) T},
+
+        with the rank :math:`R \leq n^2`. The matrices :math:`L` are diagonalized and for each
+        matrix the eigenvalues that are smaller than a given threshold (and their corresponding
+        eigenvectors) are discarded.
+
+        The algorithm has the following steps
+        [`arXiv:1902.02134 <https://arxiv.org/abs/1902.02134>`_]:
+
+        - Reshape the :math:`n \times n \times n \times n` two-electron tensor to a
+          :math:`n^2 \times n^2` matrix where :math:`n` is the number of orbitals.
+
+        - Diagonalize the resulting matrix and keep the :math:`r` eigenvectors that have
+          corresponding eigenvalues larger than a threshold.
+
+        - Multiply the eigenvectors by the square root of the eigenvalues to obtain matrices :math:`L`.
+
+        - Reshape the selected eigenvectors to :math:`n \times n` matrices.
+
+        - Diagonalize the :math:`n \times n` matrices and keep those matrices such that the norm of
+          their eigenvalues is larger than a threshold.
+    """
+    shape = two.shape
+
+    if len(shape) != 4 or len(set(shape)) != 1:
+        raise ValueError("The two-electron repulsion tensor must have a (N x N x N x N) shape.")
+
+    n = shape[0]
+    two = two.reshape(n * n, n * n)
+
+    eigvals_r, eigvecs_r = np.linalg.eigh(two)
+    eigvals_r = np.array([val for val in eigvals_r if abs(val) > tol_first])
+    eigvecs_r = eigvecs_r[:, -len(eigvals_r) :]
+
+    vectors = eigvecs_r @ np.diag(np.sqrt(eigvals_r))
+
+    r = len(eigvals_r)
+    factors = np.array([vectors.reshape(n, n, r)[:, :, k] for k in range(r)])
+
+    eigvals, eigvecs = np.linalg.eigh(factors)
+    eigvals_m = []
+    eigvecs_m = []
+    for n, eigval in enumerate(eigvals):
+        idx = [i for i, v in enumerate(eigval) if abs(v) > tol_second]
```

Comment:
Same as above comment

Reply:
An error is raised now for empty eigenvectors.